File: adapters/rest-ts/src/translation.ts
/**
 * ANIP -> REST/OpenAPI translation layer.
 *
 * Converts ANIP capability declarations into REST routes and
 * generates an OpenAPI 3.1 specification with ANIP metadata
 * preserved as x-anip-* extensions.
 */

import type { ANIPCapability, ANIPService } from "./discovery.js";
import type { RouteOverride } from "./config.js";

// Map ANIP input types to JSON Schema types
const TYPE_MAP: Record<string, string> = {
  string: "string",
  integer: "integer",
  number: "number",
  boolean: "boolean",
  date: "string",
  airport_code: "string",
};

export interface RESTRoute {
  capabilityName: string;
  path: string;
  method: string; // "GET" or "POST"
  capability: ANIPCapability;
}

export function generateRoutes(
  service: ANIPService,
  routeOverrides?: Record<string, RouteOverride>
): Map<string, RESTRoute> {
  const routes = new Map<string, RESTRoute>();
  const overrides = routeOverrides ?? {};

  for (const [name, cap] of service.capabilities) {
    let path: string;
    let method: string;

    if (name in overrides) {
      path = overrides[name].path;
      method = overrides[name].method.toUpperCase();
    } else {
      path = `/api/${name}`;
      method = cap.sideEffect === "read" ? "GET" : "POST";
    }

    routes.set(name, {
      capabilityName: name,
      path,
      method,
      capability: cap,
    });
  }

  return routes;
}

export function generateOpenAPISpec(
  service: ANIPService,
  routes: Map<string, RESTRoute>
): Record<string, unknown> {
  const paths: Record<string, unknown> = {};

  for (const [name, route] of routes) {
    const operation = buildOperation(name, route);
    const methodKey = route.method.toLowerCase();
    paths[route.path] = { [methodKey]: operation };
  }

  return {
    openapi: "3.1.0",
    info: {
      title: `ANIP REST Adapter \u2014 ${service.baseUrl}`,
      version: service.compliance,
      description:
        "Auto-generated REST API from an ANIP service. " +
        "x-anip-* extensions preserve ANIP metadata.",
    },
    paths,
    components: {
      schemas: {
        ANIPResponse: {
          type: "object",
          properties: {
            success: { type: "boolean" },
            result: { type: "object" },
            failure: {
              type: "object",
              nullable: true,
              properties: {
                type: { type: "string" },
                detail: { type: "string" },
                retry: { type: "boolean" },
                resolution: {
                  type: "object",
                  properties: {
                    action: { type: "string" },
                    requires: { type: "string" },
                    grantable_by: { type: "string" },
                  },
                },
              },
            },
            cost_actual: {
              type: "object",
              nullable: true,
              properties: {
                financial: {
                  type: "object",
                  properties: {
                    amount: { type: "number" },
                    currency: { type: "string" },
                  },
                },
                variance_from_estimate: { type: "string" },
              },
            },
            warnings: {
              type: "array",
              items: { type: "string" },
            },
          },
          required: ["success"],
        },
      },
    },
  };
}

function buildOperation(
  name: string,
  route: RESTRoute
): Record<string, unknown> {
  const cap = route.capability;

  const operation: Record<string, unknown> = {
    operationId: name,
    summary: cap.description,
    tags: ["ANIP Capabilities"],
    "x-anip-side-effect": cap.sideEffect,
    "x-anip-minimum-scope": cap.minimumScope,
    "x-anip-financial": cap.financial,
    "x-anip-contract-version": cap.contractVersion,
  };

  if (cap.cost) {
    operation["x-anip-cost"] = cap.cost;
  }
  if (cap.requires && cap.requires.length > 0) {
    operation["x-anip-requires"] = cap.requires.map((r) =>
      typeof r === "object" && r !== null ? r.capability ?? r : r
    );
  }
  if (cap.rollbackWindow) {
    operation["x-anip-rollback-window"] = cap.rollbackWindow;
  }

  // Parameters / request body
  if (route.method === "GET") {
    operation.parameters = buildQueryParameters(cap);
  } else {
    operation.requestBody = buildRequestBody(cap);
  }

  // Responses
  operation.responses = {
    "200": {
      description: "Successful ANIP response",
      content: {
        "application/json": {
          schema: { $ref: "#/components/schemas/ANIPResponse" },
        },
      },
    },
    "400": { description: "Invalid parameters" },
    "401": { description: "Delegation expired or invalid" },
    "403": {
      description:
        "Insufficient authority, budget exceeded, or purpose mismatch",
    },
    "404": { description: "Unknown capability" },
  };

  return operation;
}

function buildQueryParameters(
  cap: ANIPCapability
): Array<Record<string, unknown>> {
  const params: Array<Record<string, unknown>> = [];
  for (const inp of cap.inputs) {
    const jsonType = TYPE_MAP[inp.type ?? "string"] ?? "string";
    const param: Record<string, unknown> = {
      name: inp.name,
      in: "query",
      required: inp.required ?? true,
      schema: { type: jsonType } as Record<string, unknown>,
      description: inp.description ?? "",
    };
    if (inp.type === "date") {
      (param.schema as Record<string, unknown>).format = "date";
    }
    if (inp.default !== undefined && inp.default !== null) {
      (param.schema as Record<string, unknown>).default = inp.default;
    }
    params.push(param);
  }
  return params;
}

function buildRequestBody(cap: ANIPCapability): Record<string, unknown> {
  const properties: Record<string, unknown> = {};
  const required: string[] = [];

  for (const inp of cap.inputs) {
    const jsonType = TYPE_MAP[inp.type ?? "string"] ?? "string";
    const prop: Record<string, unknown> = {
      type: jsonType,
      description: inp.description ?? "",
    };
    if (inp.type === "date") {
      prop.format = "date";
    }
    if (inp.default !== undefined && inp.default !== null) {
      prop.default = inp.default;
    }
    properties[inp.name] = prop;
    if (inp.required ?? true) {
      required.push(inp.name);
    }
  }

  const schema: Record<string, unknown> = { type: "object", properties };
  if (required.length > 0) {
    schema.required = required;
  }

  return {
    required: true,
    content: {
      "application/json": {
        schema,
      },
    },
  };
}

File: packages/go/mcpapi/translation.go
// Package mcpapi exposes ANIP capabilities as MCP tools for stdio and HTTP transports.
package mcpapi

import (
	"fmt"
	"strings"

	"github.com/anip-protocol/anip/packages/go/core"
)

// typeMap maps ANIP types to JSON Schema types.
var typeMap = map[string]string{
	"string":       "string",
	"integer":      "integer",
	"number":       "number",
	"boolean":      "boolean",
	"date":         "string",
	"airport_code": "string",
}

// CapabilityToInputSchema converts an ANIP CapabilityDeclaration into a
// JSON Schema object suitable for MCP tool inputSchema.
func CapabilityToInputSchema(decl *core.CapabilityDeclaration) map[string]any {
	properties := make(map[string]any)
	var required []string

	for _, input := range decl.Inputs {
		jsonType, ok := typeMap[input.Type]
		if !ok {
			jsonType = "string"
		}

		prop := map[string]any{
			"type":        jsonType,
			"description": input.Description,
		}

		if input.Type == "date" {
			prop["format"] = "date"
		}

		if input.Default != nil {
			prop["default"] = input.Default
		}

		properties[input.Name] = prop

		if input.Required {
			required = append(required, input.Name)
		}
	}

	schema := map[string]any{
		"type":       "object",
		"properties": properties,
	}
	if len(required) > 0 {
		schema["required"] = required
	}

	return schema
}

// EnrichDescription builds an MCP-friendly description that includes ANIP
// metadata not representable in MCP's native schema: side-effect warnings,
// cost information, prerequisites, and scope requirements.
func EnrichDescription(decl *core.CapabilityDeclaration) string {
	parts := []string{decl.Description}

	se := decl.SideEffect
	switch se.Type {
	case "irreversible":
		parts = append(parts, "WARNING: IRREVERSIBLE action — cannot be undone.")
		if se.RollbackWindow == "none" {
			parts = append(parts, "No rollback window.")
		}
	case "write":
		if se.RollbackWindow != "" && se.RollbackWindow != "none" && se.RollbackWindow != "not_applicable" {
			parts = append(parts, fmt.Sprintf("Reversible within %s.", se.RollbackWindow))
		}
	case "read":
		parts = append(parts, "Read-only, no side effects.")
	}

	if decl.Cost != nil {
		financial := decl.Cost.Financial
		switch decl.Cost.Certainty {
		case "fixed":
			if financial != nil {
				currency := financial.Currency
				if currency == "" {
					currency = "USD"
				}
				if financial.Amount != nil {
					parts = append(parts, fmt.Sprintf("Cost: %s %v (fixed).", currency, *financial.Amount))
				}
			}
		case "estimated":
			if financial != nil {
				currency := financial.Currency
				if currency == "" {
					currency = "USD"
				}
				if financial.RangeMin != nil && financial.RangeMax != nil {
					parts = append(parts, fmt.Sprintf("Estimated cost: %s %v-%v.", currency, *financial.RangeMin, *financial.RangeMax))
				}
			}
		}
	}

	if len(decl.Requires) > 0 {
		var prereqs []string
		for _, r := range decl.Requires {
			prereqs = append(prereqs, r.Capability)
		}
		parts = append(parts, fmt.Sprintf("Requires calling first: %s.", strings.Join(prereqs, ", ")))
	}

	if len(decl.MinimumScope) > 0 {
		parts = append(parts, fmt.Sprintf("Delegation scope: %s.", strings.Join(decl.MinimumScope, ", ")))
	}

	return strings.Join(parts, " ")
}
